cmd/expiration-mailer/main.go

```diff
@@ -0,0 +1,216 @@
+
+import (
+	"bytes"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/mail"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+type emailContent struct {
+	ExpirationDate   time.Time
+	DaysToExpiration int
+	CommonName       string
+	DNSNames         string
+}
+
+type mailer struct {
+	stats         statsd.Statter
+	log           *blog.AuditLogger
+	dbMap         *gorp.DbMap
+	Mailer        mail.Mailer
+	EmailTemplate *template.Template
+	WarningDays   []int
+}
+
+func (m *mailer) findExpiringCertificates() error {
+	var err error
+	expiresStats := make(map[int]int64, len(m.WarningDays))
+	now := time.Now()
+	// E.g. m.WarningDays = [1, 3, 7, 14] days from expiration
+	for i, expiresIn := range m.WarningDays {
+		left := now
+		if i > 0 {
+			left = left.Add(time.Hour * 24 * time.Duration(m.WarningDays[i-1]))
+		}
+		right := now.Add(time.Hour * 24 * time.Duration(expiresIn))
+
+		m.log.Info(fmt.Sprintf("expiration-mailer: Searching for certificates that expire between %s and %s", left, right))
+		var certs []core.Certificate
+		_, err := m.dbMap.Select(
+			&certs,
+			`SELECT cert.* FROM certificates AS cert JOIN certificateStatus AS cs on cs.serial = cert.serial
+       WHERE cert.expires > :cutoffA AND cert.expires < :cutoffB AND cs.expirationNagsSent < :nags AND cert.status != "revoked"
+       ORDER BY cert.expires ASC`,
+			map[string]interface{}{
+				"cutoffA": left,
+				"cutoffB": right,
+				"nags":    len(m.WarningDays) - i,
+			},
+		)
+		if err != nil {
+			m.log.Err(fmt.Sprintf("expiration-mailer: Error loading certificates: %s", err))
+			continue
+		}
+		if len(certs) > 0 {
+			m.log.Info(fmt.Sprintf("expiration-mailer: Found %d certificates, starting sending messages", len(certs)))
+			for _, cert := range certs {
+				regObj, err := m.dbMap.Get(&core.Registration{}, cert.RegistrationID)
+				if err != nil {
+					return err
+				}
+				reg := regObj.(*core.Registration)
+				parsedCert, err := x509.ParseCertificate(cert.DER)
+				if err != nil {
+					return err
+				}
+				err = m.sendWarning(parsedCert, reg.Contact)
+				if err != nil {
+					return err
+				}
+				expiresStats[expiresIn]++
+
+				// Update CertificateStatus object
+				tx, err := m.dbMap.Begin()
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+
+				csObj, err := tx.Get(&core.CertificateStatus{}, cert.Serial)
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+				certStatus := csObj.(*core.CertificateStatus)
+				certStatus.ExpirationNagsSent = len(m.WarningDays) - i
```

Comment:
Storing a number of nags sent makes it difficult for us to change the number of nags. What if instead you stored a timestamp of when the last nag was sent?